scripts/template_upload_openshift.py

```diff
@@ -0,0 +1,168 @@
+normally be placed in main function, are located in function run(**kwargs).
+"""
+
+import argparse
+import os
+from threading import Lock, Thread
+
+from cfme.utils import net, trackerbot
+from cfme.utils.conf import cfme_data, credentials
+from cfme.utils.log import logger, add_stdout_handler
+from cfme.utils.providers import list_provider_keys
+from cfme.utils.ssh import SSHClient
+
+lock = Lock()
+
+add_stdout_handler(logger)
+
+
+def parse_cmd_line():
+    parser = argparse.ArgumentParser(argument_default=None)
+    parser.add_argument('--image_url', metavar='URL', dest="image_url",
+                        help="URL of OVA", default=None)
+    parser.add_argument('--template_name', dest="template_name",
+                        help="Override/Provide name of template", default=None)
+    args = parser.parse_args()
+    return args
+
+
+def check_kwargs(**kwargs):
+    for key, val in kwargs.iteritems():
+        if val is None:
+            logger.error("OPENSHIFT:%r Supply required parameter '%r'", kwargs['provider'], key)
+            return False
+    return True
+
+
+def make_kwargs(args, **kwargs):
+    args_kwargs = dict(args._get_kwargs())
+
+    if len(kwargs) is 0:
+        return args_kwargs
+
+    template_name = kwargs.get('template_name')
+    if template_name is None:
+        template_name = cfme_data['basic_info']['appliance_template']
+        kwargs.update({'template_name': template_name})
+
+    for kkey, kval in kwargs.iteritems():
+        for akey, aval in args_kwargs.iteritems():
+            if aval is not None:
+                if kkey == akey:
+                    if kval != aval:
+                        kwargs[akey] = aval
+
+    for akey, aval in args_kwargs.iteritems():
+        if akey not in kwargs.iterkeys():
+            kwargs[akey] = aval
+
+    return kwargs
+
+
+def list_templates(hostname, username, password, upload_folder):
+    with SSHClient(hostname=hostname, username=username, password=password) as ssh:
+        cmd = 'find {u}/* -maxdepth 1 -type d -exec basename {{}} \;'.format(u=upload_folder)
+        result = ssh.run_command(cmd)
+        if result.success and len(str(result)) > 0:
+            return str(result).split()
+        else:
+            return []
+
+
+def upload_template(hostname, username, password,
+                    provider, url, name, provider_data, stream, upload_folder):
+    try:
+        kwargs = {}
+
+        if name is None:
+            name = cfme_data['basic_info']['appliance_template']
+
+        logger.info("OPENSHIFT:%r Start uploading Template: %r", provider, name)
+        if not check_kwargs(**kwargs):
+            return False
+
+        if name not in list_templates(hostname, username, password, upload_folder):
+            with SSHClient(hostname=hostname, username=username, password=password) as ssh:
+                dest_dir = os.path.join(upload_folder, name)
+                result = ssh.run_command('mkdir {dir}'.format(dir=dest_dir))
```

Comment:
Add a `-p` option to mkdir to avoid common path errors.

Reply:
@mshriver, I'd prefer not to add it here since I had issues related to it before. I'd rather prefer to have error in such case .